lib/tasks/find_records_missing.rake

```diff
@@ -0,0 +1,158 @@
+    model.order(created_at: :desc).limit(args.max_per_model).map do |record|
+      if is_test_record(record)
+        owner = record.respond_to?(:user) && record.user ? record.user.email : "unknown"
+        puts "Record ##{record.id} '#{record.name}' owned by #{owner} appears to be test data"
+      end
+    end
+    puts get_total_message(args.max_per_model, model)
+  end
+end
+
+private 
+
+def is_test_record(record)
+  record.respond_to?(:name) && record.name && 
+    record.name.match(/\b(test|demo|benchmark|asdf)\b/)
+end
+
+def missing_children(model, assocs, max_per_model)
+  format_missing(model.order(created_at: :desc).limit(max_per_model).map do |record|
+    assocs
+      .map { |assoc| [record.id, assoc, record.send(assoc).pluck(:id)] }
+      .select { |data| data.last.empty? }
+      .map { |data| [data[0], data[1]] }
+  end)
+end
+
+def missing_parents(model, assocs, max_per_model)
+  format_missing(model.order(created_at: :desc).limit(max_per_model).map do |record|
+    assocs
+      .map { |assoc| [record.id, assoc, record.send(assoc)] }
+      .select { |data| data.last.nil? }
+      .map { |data| [data[0], data[1]] }
+  end)
+end
+
+def format_missing(missing)
+  missing
+  .select {|m| m.length > 0}
+  .flatten(1)
+  .group_by{|m| m[0]}
+  .map { |k, v| [k, v.map { |e| e[1] }] }
+  .to_h
+end
+
+# Run this in the rails console
+def get_initial(assoc_type=:has_many)
+  models = ApplicationRecord.descendants
+  models.map {|m| [m.name, m.reflect_on_all_associations(assoc_type).map {|a| a.name}]}
+    .select { |p| p[1].length >0 }
+    .to_h
+end
+
+# Data first obtained by get_initial(:has_many)
+def must_have_children
+  {
+    :AlignmentConfig => [:pipeline_runs],
+    :HostGenome => [:samples],
+    :MetadataField => [:metadata],
+    :PipelineRun => [
+       :pipeline_run_stages,
+       :output_states,
+       # :taxon_counts,
+       :job_stats,
+       # :taxon_byteranges,
+       # :ercc_counts,
+       # :amr_counts,
+       # :contigs
+    ],
+    # :User => %i[samples favorite_projects favorites visualizations phylo_trees],
+    :Background => [:taxon_summaries],
+    :Project => [
+       # need to use unsafe method to keep with access control
+       :samples_unsafe, 
+       # :samples, 
+       # :favorite_projects, 
+       # :favorited_by, 
+       # :phylo_trees
+    ],
+    :Sample => [
+       :pipeline_runs,
+       :input_files,
+       # :metadata
+    ]
+  }
+end
+
+# Data first obtained by get_initial(:belongs_to)
+def must_have_parents
+  {
+    :PipelineRunStage => [:pipeline_run],
+    :AmrCount => [:pipeline_run],
+    :InputFile => [:sample],
+    :OutputState => [:pipeline_run],
+    :PipelineRun => %i[sample alignment_config],
+    :TaxonCount => [:pipeline_run],
+    :Background => [:project],
+    :Contig => [:pipeline_run],
+    :ErccCount => [:pipeline_run],
+    :FavoriteProject => %i[project user],
+    :JobStat => [:pipeline_run],
+    # :Location => %i[country state subdivision city],
```

Comment:
For Locations: I can't think of any must-have relationships. They don't even have to be matched to a sample/metadatum b/c we can have higher-level Locations just for clustering (e.g. "USA" even if we don't have samples that are just at "USA")